### PR TITLE
Submit a new structure of modification

### PR DESCRIPTION
**Pull request by JoneLau**

### CODE REVIEW

File: Assets/Editor/scene.cs [hunk 1]
```diff
@@ -170,8 +171,17 @@ public class JSON_Material {
 
   [System.Serializable]
   public class JSON_Modification {
-    public string path;
-    public string name;
+    public string property;
+    public string entity;
     public object value;
+    public bool ShouldSerializeentity() {
```
Comment (jwu) [inline Assets/Editor/scene.cs]: leave a new line

File: Assets/Editor/scene.cs [hunk 2]
```diff
@@ -170,8 +171,17 @@ public class JSON_Material {
...
+      return string.IsNullOrEmpty(entity) == false;
+    }
+    public bool ShouldSerializeproperty() {
```
Comment (jwu) [inline Assets/Editor/scene.cs]: leave a new line

File: Assets/Editor/scene.cs [hunk 3]
```diff
@@ -170,8 +171,17 @@ public class JSON_Material {
...
+      return string.IsNullOrEmpty(property) == false;
+    }
+    public bool ShouldSerializevalue() {
```
Comment (jwu) [inline Assets/Editor/scene.cs]: leave a new line

File: Assets/Editor/Exporter.Scene.cs [hunk 4]
```diff
@@ -77,7 +77,7 @@ public partial class Exporter {
         result.scale[1] = _go.transform.localScale.y;
         result.scale[2] = _go.transform.localScale.z;
       }
-
+      result.index=_nodes.IndexOf(_go);
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]: remove this (no need)

File: Assets/Editor/scene.cs [hunk 5]
```diff
@@ -32,6 +32,7 @@ public class JSON_Entity {
     // basic
     public string name;
     public string prefab;
+    public int index;
```
Comment (jwu) [inline Assets/Editor/scene.cs]: remove this (no need)

File: Assets/Editor/Exporter.Scene.cs [hunk 6]
```diff
@@ -93,27 +93,51 @@ public partial class Exporter {
...
-        //         foreach (Material mat in renderer.sharedMaterials) {
-        //           id = Utils.AssetID(mat);
-        //           matAssets.Add(id);
-        //         }
-        //       }
-        //     }
-        //   }
-        // }
+        PropertyModification[] mods = PrefabUtility.GetPropertyModifications(_go);
+        if (mods.Length > 0) {
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]:
```
if (mods.length == 0) {
  continue;
}
```

File: Assets/Editor/Exporter.Scene.cs [hunk 7]
```diff
@@ -93,27 +93,51 @@ public partial class Exporter {
...
+          foreach (var mod in mods) {
+            JSON_Modification modification = null;
+            if (mod.target is MeshRenderer) {
+              var renderer = mod.target as MeshRenderer; 
+              if (mod.propertyPath.Contains("Materials")) {
+                modification = new JSON_Modification();
+                string type = Utils.modificationTypeInfo.TryGetValue(renderer.GetType().ToString(), out type) ? type + "." : "";
+                modification.property = type + mod.propertyPath;
+                GameObject gameSelf = renderer.gameObject;
+                modification.entity = "nodes.IndexOf(" + nodes.IndexOf(gameSelf) + ")";
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]: modification.entity = nodes.IndexOf(go);

File: Assets/Editor/Exporter.Scene.cs [hunk 8]
```diff
@@ -93,27 +93,51 @@ public partial class Exporter {
...
+          });
+          foreach (var mod in mods) {
+            JSON_Modification modification = null;
+            if (mod.target is MeshRenderer) {
+              var renderer = mod.target as MeshRenderer; 
+              if (mod.propertyPath.Contains("Materials")) {
+                modification = new JSON_Modification();
+                string type = Utils.modificationTypeInfo.TryGetValue(renderer.GetType().ToString(), out type) ? type + "." : "";
+                modification.property = type + mod.propertyPath;
+                GameObject gameSelf = renderer.gameObject;
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]: gameSelf change to go

**Comment by JoneLau** [comment]
update cocos-creator/engine-3d#82

File: Assets/Editor/Exporter.Scene.cs [hunk 9]
```diff
@@ -77,7 +77,7 @@ public partial class Exporter {
         result.scale[1] = _go.transform.localScale.y;
         result.scale[2] = _go.transform.localScale.z;
       }
-
+      //result.index=_nodes.IndexOf(_go);
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]: remove the line please :)

File: Assets/Editor/Exporter.Scene.cs [hunk 10]
```diff
@@ -94,26 +94,44 @@ public partial class Exporter {
...
-        //         List<string> matAssets = new List<string>();
-        //         foreach (Material mat in renderer.sharedMaterials) {
-        //           id = Utils.AssetID(mat);
-        //           matAssets.Add(id);
-        //         }
-        //       }
-        //     }
-        //   }
-        // }
+        // TODO:
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]: remove the line please

File: Assets/Editor/Exporter.Scene.cs [hunk 11]
```diff
@@ -94,26 +94,44 @@ public partial class Exporter {
...
-        //         foreach (Material mat in renderer.sharedMaterials) {
-        //           id = Utils.AssetID(mat);
-        //           matAssets.Add(id);
-        //         }
-        //       }
-        //     }
-        //   }
-        // }
+        // TODO: 
+        // // Dump Modifications
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]: remove unused `//`

File: Assets/Editor/scene.cs [hunk 12]
```diff
@@ -32,6 +32,7 @@ public class JSON_Entity {
     // basic
     public string name;
     public string prefab;
+    //public int index;
```
Comment (jwu) [inline Assets/Editor/scene.cs]: remove the line please

File: Assets/Editor/Exporter.Scene.cs [hunk 13]
```diff
@@ -94,26 +94,44 @@ public partial class Exporter {
...
+          foreach (var mod in mods) {
+            JSON_Modification modification = null;
+            if (mod.target is MeshRenderer) {
+              var renderer = mod.target as MeshRenderer; 
+              if (mod.propertyPath.Contains("Materials")) {
+                modification = new JSON_Modification();
+                string type = Utils.modificationTypeInfo.TryGetValue(renderer.GetType().ToString(), out type) ? type + "." : "";
+                modification.property = type + mod.propertyPath;
+                GameObject go = renderer.gameObject;
+                modification.entity = "nodes.IndexOf(" + nodes.IndexOf(go) + ")";
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]: `modification.entity = nodes.IndexOf(go);`

File: Assets/Editor/scene.cs [hunk 14]
```diff
@@ -170,8 +171,20 @@ public class JSON_Material {
 
   [System.Serializable]
   public class JSON_Modification {
-    public string path;
-    public string name;
+    public string property;
+    public string entity;
```
Comment (jwu) [inline Assets/Editor/scene.cs]: public int entity;

File: Assets/Editor/scene.cs [hunk 15]
```diff
@@ -170,8 +171,20 @@ public class JSON_Material {
...
     public object value;
+
+    public bool ShouldSerializeentity() {
+      return string.IsNullOrEmpty(entity) == false;
```
Comment (jwu) [inline Assets/Editor/scene.cs]: return entity != -1;

**Pull request merged**